sys/net/application_layer/nanocoap/nanocoap.c

```diff
@@ -406,6 +406,52 @@ ssize_t coap_opt_get_string(coap_pkt_t *pkt, uint16_t optnum,
     return (int)(max_len - left);
 }
 
+int coap_iterate_uri_query(coap_pkt_t *pkt, void **opt_pos,
+                           char *key, size_t key_len_max,
+                           char *value, size_t value_len_max)
+{
+    int len;
+    void *data = coap_iterate_option(pkt, COAP_OPT_URI_QUERY, (uint8_t **)opt_pos, &len);
+    if (!data) {
+        return 0;
+    }
+
+    const char *sep = memchr(data, '=', len);
+
+    size_t key_len;
+    const char *val_data;
+    size_t val_len;
+
+    if (sep) {
+        key_len = (uintptr_t)sep - (uintptr_t)data;
+        val_data = sep + 1;
+        val_len = len - key_len - 1;
```

Comment:
Can we do a special error return value in case of malformed query parameters? For example with `?key1=&key2=2` `key1` will lead to `val_len = 0`.

Reply:
That is not malformed, that is perfectly within spec.

Reply:
`?key=7&` as well?

Reply:
Yes.

The `=` sign is merely convention to start with, and any Uri-Path option that contains a UTF-8 string (with caveats that are too niche to matter) is valid.

Also, a single value `["foo&bar"]`, which when turned into a URI reference would become `?foo%26bar`, and thus distinct from `?foo&bar` which is `["foo", "bar"]`. Which is one of the reasons I'm not a friend of using URI parsing on CoAP options -- not only was it specifically designed that you don't need to, it's also very easy to get wrong.